File: 136-Single-Number.py
# Problem: Given an array of duplicates and a single number
# Return: The single number

class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        counter = dict()
        for elm in nums:
            if elm in counter:
                counter[elm] += 1
            else:
                counter[elm] = 1
        return([k for k,v in counter.items() if v == 1][0])
# -----------------------------------------------------------------------
driver = Solution()

t0 = [0,0,1,1,2,2,3,3,5]
t1 = [2,3,1,1,4,3,2]

print('\n')
print('Array:',t0,'Program vs Expected sol:\n',driver.singleNumber(t0),'\n 5\n')
print('Array:',t1,'Program vs Expected sol:\n',driver.singleNumber(t1),'\n 4\n')

File: 136-Single-Number.cpp
#include <iostream>
#include <vector>

// Problem:
// Given a non-empty array of integers, every element appears twice except for one. 
// Find that single one.

class Solution {
public:
    int singleNumber(std::vector<int>& nums) {
        int single_number = nums[0];
        for(auto num_iter = nums.begin()+1; num_iter != nums.end(); num_iter++)
        {
            single_number ^= *num_iter;
        }
        return single_number;
    }
};